src/corelib/ui/widgets/pie/pie.cpp

```diff
@@ -0,0 +1,371 @@
+    sliceOpacityAnimation.start();
+}
+
+void Pie::Implementation::animateDeselectedSlice()
+{
+    sliceOpacityAnimation.setCurrentTime(0);
+    sliceOpacityAnimation.setDirection(QVariantAnimation::Backward);
+    sliceOpacityAnimation.start();
+}
+
+Pie::Pie(QWidget* _parent)
+    : Widget(_parent)
+    , d(new Implementation)
+{
+    this->setMouseTracking(true);
+
+    connect(&d->sliceOpacityAnimation, &QVariantAnimation::valueChanged, this,
+            [this] { update(); });
+}
+
+Pie::Pie(const QAbstractItemModel* model, const int valueColumn, QWidget* _parent)
+    : Widget(_parent)
+    , d(new Implementation(model, valueColumn))
+{
+    this->setMouseTracking(true);
+
+    connectSignals(model);
+
+    connect(&d->sliceOpacityAnimation, &QVariantAnimation::valueChanged, this,
+            [this] { update(); });
+}
+
+Pie::~Pie()
+{
+}
+
+void Pie::paintEvent(QPaintEvent* _event)
+{
+    Q_UNUSED(_event)
+
+    QPainter painter(this);
+    painter.setRenderHint(QPainter::Antialiasing);
+
+    if (d->sliceOpacityAnimation.state() == QVariantAnimation::Running) {
+        painter.setOpacity(d->sliceOpacityAnimation.currentValue().toReal());
+    } else if (d->selectedSlice) {
+        painter.setOpacity(d->sliceOpacityAnimation.endValue().toReal());
+    }
+
+    for (const auto& slice : d->slices) {
+        if (d->lastSelectedSlice == slice.get()) {
+            continue;
+        }
+        painter.fillPath(slice->path, slice->color);
+    }
+
+    if (d->lastSelectedSlice != nullptr) {
+        painter.setOpacity(1.0);
+        painter.fillPath(d->lastSelectedSlice->path, d->lastSelectedSlice->color);
+    }
+}
+
+void Pie::resizeEvent(QResizeEvent* _event)
+{
+    Q_UNUSED(_event);
+    recalculateSlices();
+}
+
+void Pie::mouseMoveEvent(QMouseEvent* _event)
+{
+    for (const auto& slice : d->slices) {
+        if (slice->path.contains(_event->pos())) {
+            if (d->selectedSlice != slice.get()) {
+                if (d->selectedSlice == nullptr) {
+                    d->animateSelectedSlice();
+                } else {
+                    update();
+                }
+
+                d->selectedSlice = slice.get();
+                d->lastSelectedSlice = d->selectedSlice;
+
+                emit itemSelected(slice->index);
+                return;
+            } else {
+                return;
+            }
+        }
+    }
+
+    if (d->selectedSlice != nullptr) {
+        d->selectedSlice = nullptr;
+        d->animateDeselectedSlice();
+    }
+}
+
+void Pie::recalculateSlices()
+{
+    auto min = std::min(size().width(), size().height());
+    d->recalculateSlices(
+        { { (rect().width() - min) / 2.0, (rect().height() - min) / 2.0 }, QSize{ min, min } });
+
+    update();
+}
+
+void Pie::insertSlices(const QModelIndex& parent, int first, int last)
+{
+    Q_UNUSED(parent)
+
+    auto iter = d->slices.begin();
+    std::advance(iter, first);
+
+    for (int i = first; i <= last; i++) {
```

Comment:
лучше префиксный инкремент использовать

Reply:
а компилятор разве такие вещи сам не исправляет?

Reply:
ни разу о таком не слышал доселе, понятно, что это экономия на спичках, но best practice, как ни как :)

Reply:
Одинаковый код
![image](https://user-images.githubusercontent.com/67420634/124460452-72797380-dda0-11eb-84d1-fa930875e4ed.png)

Reply:
> The prefix version avoids unnecessary work. It increments the value and returns the incremented version.The postfix operator must store the original value so that it can return the unincremented value as its result. If we don’t need the unincremented value, there’s no need for the extra work done by the postfix operator.
> 
> For ints and pointers, the compiler can optimize away this extra work. For more complicated iterator types, this extra work potentially might be more costly.

Reply:
ну так, что меняю везде на префиксный? А тож вроде как в этой ситуации код будет одинаковый

Reply:
давай поменяем, плюс будет консистентно по всему коду у нас

Reply:
исправил